### PR TITLE
c-deps: vendor with submodules, not tarballs

### PR DESCRIPTION
**Pull request by benesch**
Since #14840 (4d72e12), we've vendored our C and C++ dependencies as version-controlled tarballs. This avoided a dependency on autotools: upstream would run autotools to produce configure from configure.ac, Makefile from Makefile.am, etc., and package those generated files into the tarballs.

These tarballs caused quite a few problems, though. Maintaining patchsets required a homegrown shell script, import.sh, that had none of Git's conflict resolution features when patches became outdated. Applying these patches caused the tarball to depend on the modification time of the affected files, meaning the tarball had a different hash every time import.sh was run. Most importantly, a PR slipped through that claimed to upgrade to v5.5.1 of RocksDB while shipping a tarball built from a different commit (#16808), since the tarballs are opaque binary blobs.

For the other binary blobs that are checked in to version control, like embedded.go, CI can verify their contents by regenerating them from their sources and diffing. Since the tarballs are not reproducible, we can't easily teach CI to verify them.

Instead, this commit swaps out the tarballs for Git submodules. This solves all of the problems above at the cost of introducing a dependency on autotools. Reviewers now have total visibility into the dependency under review, thanks to GitHub's submodule UI, and maintaining patchsets is as easy as maintaining a Git branch on the cockroachdb-owned fork of the upstream repository.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -18,6 +18,11 @@ required = [
  "golang.org/x/tools/cmd/stringer",
 ]
 
+ignored = [
+  # Non-existent package used by a toy program in c-deps/protobuf.
+  "github.com/google/protobuf/examples/tutorial"
```
Comment (dt) [inline Gopkg.toml]: 🥇

**Comment by tamird** [comment]
:lgtm:

I took a quick look at https://github.com/cockroachdb/rocksdb/branches - before you merge this, would you mind pruning out unused branches from these new forks?

---

Reviewed 20 of 20 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[scripts/ls-files.sh, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoIxYJ-odWgW28vMqgt:-KoIxYJ-odWgW28vMqgu:bul60s4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bfbd6cc9da6c61f8f78b7d3566a82029cf1f8b5/scripts/ls-files.sh#L19)):*
> ```Shell
> git ls-files . ':!vendor' ':!c-deps'
> for submodule in vendor c-deps/*
> do
> ```

nit: i think we usually write `for ...; do`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-:-KoIxFeJ7NvKtzHmROoS:b-druk3z)*

**Comment by bdarnell** [comment]
:lgtm:

---

Reviewed 19 of 20 files at r1, 4 of 4 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[CONTRIBUTING.md, line 15 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoJ5_I1tK3fA5nevuww:-KoJ5_I1tK3fA5nevuwx:b1uo1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/388e27b6c4f94a41c7609c236213747fb17acdda/CONTRIBUTING.md#L15)):*
> ```Markdown
>   - GNU Make (3.81+ is known to work)
>   - CMake 2.8.12+
>   - Autotools 2.68+
> ```

Should we list autoconf, automake, and libtool separately? I see that's how we install them in the dockerfile. We should try to match the names used in the apt/brew/etc repositories as much as possible in this list. 

---

*[CONTRIBUTING.md, line 16 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoJ3BaeVHYJWAS74LC9:-KoJ3BaeVHYJWAS74LCA:bgks4gm) ([raw file](https://github.com/cockroachdb/cockroach/blob/388e27b6c4f94a41c7609c236213747fb17acdda/CONTRIBUTING.md#L16)):*
> ```Markdown
>   - CMake 2.8.12+
>   - Autotools 2.68+
>   - XZ Utils (5.2.3+ is known to work), except on macOS, where xz support is
> ```

We no longer need `xz-utils`, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-:-KoJ6AyXsy125JMFkhy2:bnfp4nl)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[build/Dockerfile, line 20 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoJ6W5ZG7h8LRMzg2u4:-KoJ6W5ZG7h8LRMzg2u5:b-rudcvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/074a80f9c2a2227b4b32b5d087a4a3a998fa9e74/build/Dockerfile#L20)):*
> ```Dockerfile
>  && echo 'deb https://dl.yarnpkg.com/debian/ stable main' | tee /etc/apt/sources.list.d/yarn.list
> 
> # autoconf - crosstool-ng/bootstrap / CRDB build system
> ```

consider being more specific with these, perhaps by listing which c-deps need them

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859)*

**Comment by benesch** [comment]
Unused branches are pruned. Updating the builder image with automake/libtool now!

---

Review status: 23 of 26 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[CONTRIBUTING.md, line 15 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoJ5_I1tK3fA5nevuww:-KoJBP71wPFEcgzlAEYM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/388e27b6c4f94a41c7609c236213747fb17acdda/CONTRIBUTING.md#L15)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Should we list autoconf, automake, and libtool separately? I see that's how we install them in the dockerfile. We should try to match the names used in the apt/brew/etc repositories as much as possible in this list. 
</blockquote></details>

Done.

---

*[CONTRIBUTING.md, line 16 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoJ3BaeVHYJWAS74LC9:-KoJBPVHTNxGqIB9o3Ri:bombmv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/388e27b6c4f94a41c7609c236213747fb17acdda/CONTRIBUTING.md#L16)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We no longer need `xz-utils`, right?
</blockquote></details>

Good call. Done.

---

*[Gopkg.toml, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoIvlEqyPwv-j3nJkgI-r1-23:-KoJBc2UGYMb4r9HNqNj:bnacbln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bfbd6cc9da6c61f8f78b7d3566a82029cf1f8b5/Gopkg.toml#L23)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

🥇 
</blockquote></details>

:bow: 

---

*[build/Dockerfile, line 20 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoJ6W5ZG7h8LRMzg2u4:-KoJBRXsBkrwhtXctKhk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/074a80f9c2a2227b4b32b5d087a4a3a998fa9e74/build/Dockerfile#L20)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

consider being more specific with these, perhaps by listing which c-deps need them
</blockquote></details>

Done.

---

*[scripts/ls-files.sh, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoIxYJ-odWgW28vMqgt:-KoJBSGZn28QA2j7hfLl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bfbd6cc9da6c61f8f78b7d3566a82029cf1f8b5/scripts/ls-files.sh#L19)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: i think we usually write `for ...; do`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-:-KoJBUCWUy7ymIwHtPBN:b-1zusdm)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[scripts/ls-files.sh, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoIxYJ-odWgW28vMqgt:-KoJCvM0Cu7MhKp2T66j:bbo7m8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bfbd6cc9da6c61f8f78b7d3566a82029cf1f8b5/scripts/ls-files.sh#L19)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Does the comment above need an update? I see you added a new comment, but it's no longer clear to me how everything fits together.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859)*

**Comment by benesch** [comment]
Review status: 25 of 26 files reviewed at latest revision, 4 unresolved discussions.

---

*[scripts/ls-files.sh, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoIxYJ-odWgW28vMqgt:-KoJET9ZLZA58wm_dYbH:b-fo91js) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bfbd6cc9da6c61f8f78b7d3566a82029cf1f8b5/scripts/ls-files.sh#L19)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Does the comment above need an update? I see you added a new comment, but it's no longer clear to me how everything fits together.
</blockquote></details>

Strictly speaking, no update necessary, but I can see how it might be confusing now. I tweaked the organization a tad to make it clearer. Take another peek?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[scripts/ls-files.sh, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoIxYJ-odWgW28vMqgt:-KoJExBd_ExJ-TvjTran:b-iao4ql) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bfbd6cc9da6c61f8f78b7d3566a82029cf1f8b5/scripts/ls-files.sh#L19)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Strictly speaking, no update necessary, but I can see how it might be confusing now. I tweaked the organization a tad to make it clearer. Take another peek?
</blockquote></details>

Yes, that's clearer. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859)*

**Comment by tamird** [comment]
Reviewed 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[c-deps/snappy, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoJi5ouSh3iEUn2JeCU:-KoJi5ouSh3iEUn2JeCV:bj7g1nw) ([raw file](https://github.com/cockroachdb/cockroach/blob/14d6a45662c59488c6da9dcbaf030cd9a05bf9a7/c-deps/snappy#L1)):*
> ```
> Subproject commit 15fb19fc140bcd5e2c529c99a6d87a0fe2de1a8e
> ```

What changed here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859)*

**Comment by benesch** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[c-deps/snappy, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoJi5ouSh3iEUn2JeCU:-KoKVp8Y6UYR_3oRUvXm:botn7ir) ([raw file](https://github.com/cockroachdb/cockroach/blob/14d6a45662c59488c6da9dcbaf030cd9a05bf9a7/c-deps/snappy#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

What changed here?
</blockquote></details>

Build system golf: https://github.com/cockroachdb/snappy/commit/15fb19fc140bcd5e2c529c99a6d87a0fe2de1a8e

A dependency on pkg-config for one disused configure option seemed real unfortunate. Snappy switched to CMake in 1.1.5, so I have a better solution coming down the pipe.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[c-deps/snappy, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859#-KoJi5ouSh3iEUn2JeCU:-KoKWdwPOVvuzqNIyCVh:bnjzbbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/14d6a45662c59488c6da9dcbaf030cd9a05bf9a7/c-deps/snappy#L1)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Build system golf: https://github.com/cockroachdb/snappy/commit/15fb19fc140bcd5e2c529c99a6d87a0fe2de1a8e

A dependency on pkg-config for one disused configure option seemed real unfortunate. Snappy switched to CMake in 1.1.5, so I have a better solution coming down the pipe.
</blockquote></details>

:tada:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16859)*

**Pull request merged**